File: src/ThemeParkPlanner.Console/Attraction.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Challenge.Core;

namespace ThemeParkPlanner
{
    public class Attraction
    {
        /// <summary>
        /// Gets the QueueTimes indexed by the hour in which the time occurs.
        /// </summary>
        public IReadOnlyDictionary<int, int> QueueTimes { get; private set; }

        public Attraction(IEnumerable<int> queueTimes)
        {
            // ReSharper disable PossibleMultipleEnumeration
            QueueTimes = Enumerable.Range(0, queueTimes.Count())
                .ToDictionary(i => i, queueTimes.ElementAt);
        }

        public bool TryGetQueueTime(int timeMinutes, out int queueTime)
        {
            var hour = timeMinutes/Constants.MinutesPerHour;
            return QueueTimes.TryGetValue(hour, out queueTime);
        }

        private static Attraction ReadOne(TextReader reader)
        {
            var queueTimes = from x in reader.ReadLineAsync().Result.Split(' ')
                select x.ParseInteger();
            return new Attraction(queueTimes);
        }

        public static IEnumerable<Attraction> ReadAll(TextReader reader, int count)
        {
            while (count-- > 0)
                yield return ReadOne(reader);
        }
    }
}

File: src/ParkPlannerCpp/attraction.cpp
#include "attraction.h"
#include "constants.h"

#include <string>

#include <common\parsers.h>
#include <common\strutil.h>

namespace challenges {

    namespace park {

        attraction::attraction(std::vector<int> const & queue_times)
            : _queue_times()
        {
            for (auto i = 0; i < static_cast<int>(queue_times.size()); i++)
                _queue_times.emplace(i, *(queue_times.begin() + i));
        }

        attraction::attraction(attraction const & other)
            : _queue_times(other._queue_times){
        }

        attraction::~attraction()
        {
            _queue_times.clear();
        }

        bool attraction::try_get_queue_time(int time_minutes, int & queue_time) {
            queue_time = 0;
            auto hour = time_minutes / constants::minutes_per_hour;
            if (_queue_times.find(hour) == _queue_times.end()) return false;
            queue_time = _queue_times.at(hour);
            return true;
        }

        attraction attraction::read_one(std::istream & is) {

            std::string line;

            getline(is, line);

            split splitter;

            std::vector<int> queue_times;

            for (auto const & x : splitter(line))
                queue_times.push_back(to_int(x));

            return attraction(queue_times);
        }

        std::vector<attraction> attraction::read_all(std::istream & is, int count) {

            std::vector<attraction> results;

            while (count-- > 0)
                results.push_back(read_one(is));

            return results;
        }
    }
}